lib/reek/smells/subclassed_from_core_class.rb

```diff
@@ -0,0 +1,64 @@
+# frozen_string_literal: true
+require_relative 'smell_detector'
+require_relative 'smell_warning'
+
+module Reek
+  module Smells
+    #
+    # Subclassing core classes in Ruby can lead to unexpected side effects.
+    # Knowing that Ruby has a core library, which is written in C, and a standard
+    # library, which is written in Ruby, if you do not know exactly how these core
+    # classes operate at the C level, you are gonna have a bad time.
+    #
+    # Source: http://words.steveklabnik.com/beware-subclassing-ruby-core-classes
+    #
+    class SubclassedFromCoreClass < SmellDetector
+      CORE_CLASSES = ['Array', 'Hash', 'String'].freeze
+
+      METHODS = {
+        casgn: :inspect_casgn,
+        class: :inspect_class
+      }.freeze
+
+      def self.contexts
+        [:class, :casgn]
+      end
+
+      # Checks +ctx+ for if it is subclasssed from a core class
+      #
+      # @return [Array<SmellWarning>]
+      def inspect(ctx)
+        public_send(METHODS[ctx.type], ctx) || []
```

Comment:
I believe that's a tad over optimized here. inspect_class and inspect_casgn are almost identical and contain some duplication.
How about dissolving both of them and handle everything in "inspect"? I dont think this will blow up "inspects" method size and i believe it will be clearer in the end.